### ISSUE TITLE
Failure to set memory limits in cgroups

### ISSUE THREAD

**Issue by wpoely86** [adaptivecomputing/torque#341]
We have an issue with the cgroup creation. This was done on 37ae8750b2e
The jobscript is
```bash
#!/bin/bash

#PBS -q long
#PBS -l walltime=11:59:00
#PBS -l nodes=4:ppn=1

sleep 72h
```
The job will start but return immediately with code -3 (`JOB_EXEC_RETRY`) and will keep on retrying. This is caused by a failure to create the cgroups. In the node logs we find:
```
11/20/2015 12:07:14.472;01;   pbs_mom.1480;Svr;pbs_mom;LOG_DEBUG::init_groups, pre-sigprocmask
11/20/2015 12:07:14.473;01;   pbs_mom.1480;Svr;pbs_mom;LOG_DEBUG::init_groups, post-initgroups
11/20/2015 12:07:14.476;01;   pbs_mom.1480;Svr;pbs_mom;LOG_ERROR::trq_cg_create_task_cgroups, Failed to get allocation for task_index 2: error 15004
11/20/2015 12:07:14.476;01;   pbs_mom.1480;Svr;pbs_mom;LOG_ERROR::trq_cg_create_cgroup, failed to create task cgroups: 15004
```
We can find this error in all nodes that are used for the job. The OS is CentOS 7. A single node job works without issue.

cc @stdweird

**Comment by wpoely86** [comment]
To be clear, the test nodes we use have a `ppn=2`. So `nodes=4:ppn=1` means 2 nodes. If I changes this to `nodes=2:ppn=1` (which only needs a single node), it also works. Only when the job is spread over multiple nodes, the cgroup error appears.

**Comment by knielson** [comment]
I will see if we can reproduce the behavior here. If single node jobs are
working it sounds like the cgroup file system is mounted.

Ken

On Fri, Nov 20, 2015 at 5:58 AM, wpoely86 <notifications@github.com> wrote:

> To be clear, the test nodes we use have a ppn=2. So nodes=4:ppn=1 means 2
> nodes. If I changes this to nodes=2:ppn=1 (which only needs a single
> node), it also works. Only when the job is spread over multiple nodes, the
> cgroup error appears.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/adaptivecomputing/torque/issues/341#issuecomment-158392361>
> .
>



-- 
[image: Adaptive Computing] <http://www.adaptivecomputing.com>
[image: Twitter] <http://twitter.com/AdaptiveMoab> [image: LinkedIn]
<http://www.linkedin.com/company/448673?goback=.fcs_GLHD_adaptive+computing_false_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2&trk=ncsrch_hits>
[image:
YouTube] <http://www.youtube.com/adaptivecomputing> [image: GooglePlus]
<https://plus.google.com/u/0/102155039310685515037/posts> [image: Facebook]
<http://www.facebook.com/pages/Adaptive-Computing/314449798572695?fref=ts>
[image:
RSS] <http://www.adaptivecomputing.com/feed>
Ken Nielson Sr. Software Engineer
+1 801.717.3700 office    +1 801.717.3738 fax
1712 S. East Bay Blvd, Suite 300     Provo, UT 84606
www.adaptivecomputing.com

**Comment by wpoely86** [comment]
@knielson This is on a CentOS 7 system. So cgroups are definitely mounted (or systemd won't work). For single node jobs, I can see the created cgroups and the pid's of the job are part of them. The memory cgroup however does not limit anything, even if `vmem` is explicitly set in the job file.

**Comment by knielson** [comment]
The memory limit problem is a known bug. I am working on that right now.

Ken

On Mon, Nov 23, 2015 at 12:41 PM, wpoely86 <notifications@github.com> wrote:

> @knielson <https://github.com/knielson> This is on a CentOS 7 system. So
> cgroups are definitely mounted (or systemd won't work). For single node
> jobs, I can see the created cgroups and the pid's of the job are part of
> them. The memory cgroup however does not limit anything, even if vmem is
> explicitly set in the job file.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/adaptivecomputing/torque/issues/341#issuecomment-159041123>
> .
>



-- 
[image: Adaptive Computing] <http://www.adaptivecomputing.com>
[image: Twitter] <http://twitter.com/AdaptiveMoab> [image: LinkedIn]
<http://www.linkedin.com/company/448673?goback=.fcs_GLHD_adaptive+computing_false_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2_*2&trk=ncsrch_hits>
[image:
YouTube] <http://www.youtube.com/adaptivecomputing> [image: GooglePlus]
<https://plus.google.com/u/0/102155039310685515037/posts> [image: Facebook]
<http://www.facebook.com/pages/Adaptive-Computing/314449798572695?fref=ts>
[image:
RSS] <http://www.adaptivecomputing.com/feed>
Ken Nielson Sr. Software Engineer
+1 801.717.3700 office    +1 801.717.3738 fax
1712 S. East Bay Blvd, Suite 300     Provo, UT 84606
www.adaptivecomputing.com

**Comment by dbeer** [comment]
This should be resolved now on the latest 6.0-dev. There are multiple commits involved, but can you re-test with 6.0-dev?

**Comment by itkovian** [comment]
The memory is still not set correctly in the cgroups.

**Comment by stdweird** [comment]
to clarify, we are seeing the following: 
* server uses `legcay_vmem  = true`
* jobs are submitted with `-l vmem=... -l nodes=2:ppn=2`

The cgroup memory limit of the job is determined via https://github.com/adaptivecomputing/torque/blob/6.0-dev/src/resmom/start_exec.c#L4278 (and not using a complete_req), and the returned value is not in line with our understanding of how legacy_vmem should work

The swap/vmem limit is not set; in particular, there is no equivalent function https://github.com/adaptivecomputing/torque/blob/6.0-dev/src/resmom/start_exec.c#L4278 that deals with vmem/swap

**Comment by itkovian** [comment]
Latest 6.0-dev, commit 24827b8

~~~~
vsc40075@node2803 (banette) ~> qsub -I -l walltime=2:00:00:00 -l nodes=2:ppn=2 -l vmem=8g
qsub: waiting for job 340.master23.banette.gent.vsc to start
qsub: job 340.master23.banette.gent.vsc ready

vsc40075@node2801 (banette) ~> cat /sys/fs/cgroup/memory/torque/340.master23.banette.gent.vsc/memory.limit_in_bytes
4294967296
vsc40075@node2801 (banette) ~> cat /sys/fs/cgroup/memory/torque/340.master23.banette.gent.vsc/memory.memsw.limit_in_bytes
9223372036854775807
~~~~

So the limit_in_bytes is still set to half of what is expected under the legacy_vmem=True setting.

**Comment by stdweird** [comment]
@dbeer @itkovian to be clear, there are *2* things wrong
* the mem limit is set to only half of what is expected: we expect it to be set to the vmem value (and the vmem value should the requested value with legacy_vmem enabled), and the only reason being that you cannot set the memsw value without setting the mem value
* the memsw limit is not set at all, even if we ask `vmem` (and we expect it to be set to the vmem value as this is with legacy_vmem enabled)

**Comment by itkovian** [comment]
Last update with torque-6.0-ug.1466177621.6148df8:

~~~~
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/memory.kmem.limit_in_bytes:9223372036854775807
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/memory.kmem.tcp.limit_in_bytes:9223372036854775807
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/memory.limit_in_bytes:1610612736
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/memory.memsw.limit_in_bytes:9223372036854775807
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/memory.soft_limit_in_bytes:9223372036854775807
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/tasks:24813
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/tasks:24814
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/tasks:24826
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/tasks:24827
/sys/fs/cgroup/memory/torque/346.master23.banette.gent.vsc/tasks:24828
~~~~

~~~~
    Resource_List.nice = 0
    Resource_List.nodect = 2
    Resource_List.nodes = 2:ppn=2
    Resource_List.vmem = 3gb
    Resource_List.walltime = 00:10:00
~~~~

**Comment by itkovian** [comment]
Same thing for commit 332175e.

**Comment by itkovian** [comment]
See #372, both memory.memsw.limit_in_bytes and memory.limit_in_bytes are set to the expected value under legacy_vmem=true using #378.

**Comment by itkovian** [comment]
Note that c3999fd04 does not fix this issue. Please check the locations alias_gethostname is used in #378

**Comment by itkovian** [comment]
This seems to be fixed under 56698fb.
